backend/kafka_backend.go

```diff
@@ -0,0 +1,114 @@
+package backend
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"sync"
+
+	"github.com/confluentinc/confluent-kafka-go/kafka"
+	"github.com/skroutz/downloader/job"
+)
+
+var eventsWG sync.WaitGroup
+
+// KafkaBackend notifies about a job completion by producing to a Kafka topic.
+type KafkaBackend struct {
+	producer *kafka.Producer
+	reports  chan job.CallbackInfo
+}
+
+// ID returns the identifier code for the respective backend
+// which is "kafka".
+func (b *KafkaBackend) ID() string {
+	return "kafka"
+}
+
+// Start sets up a kafka producer and a channel for delivery reports.
+func (b *KafkaBackend) Start(ctx context.Context, cfg map[string]interface{}) error {
+	var (
+		err         error
+		kafkaConfig kafka.ConfigMap
+	)
+
+	kafkaConfig = make(kafka.ConfigMap)
+
+	for key, value := range cfg {
+		kafkaConfig.SetKey(key, value)
+	}
+
+	b.producer, err = kafka.NewProducer(&kafkaConfig)
+	if err != nil {
+		return err
+	}
+
+	b.reports = make(chan job.CallbackInfo)
+
+	// start a go routine to monitor Kafka's Events channel
+	go b.transformStream(ctx, &eventsWG)
+
+	return nil
+}
+
+// Notify produces (asynchronously) a payload message into a kafka topic using
+// the producer. Kafka will send the delivery report to the default Events delivery channel.
+func (b *KafkaBackend) Notify(destination string, payload []byte) error {
+	message := &kafka.Message{
+		TopicPartition: kafka.TopicPartition{Topic: &destination, Partition: kafka.PartitionAny},
+		Value:          payload,
+	}
+	err := b.producer.Produce(message, nil)
+	return err
+}
+
+// DeliveryReports returns a channel of emmited callback info events
+func (b *KafkaBackend) DeliveryReports() <-chan job.CallbackInfo {
+	return b.reports
+}
+
+// Finalize closes the reports channel and the producer.
+// It also handles gracefully the termination of any go routines
+// that the kafka backend started.
+func (b *KafkaBackend) Finalize() error {
+	close(b.reports)
+	eventsWG.Wait()
+	b.producer.Close()
+	return nil
+}
+
+// transformStream iterates over the Events channel of Kafka, transforms
+// each message to a callback info object and enqueues the callback info object
+// to our reports channel.
+func (b *KafkaBackend) transformStream(ctx context.Context, wg *sync.WaitGroup) {
+	wg.Add(1)
```

Comment:
It's important that `wg.Add()` is called _before_ the goroutine is spawned.

> Note that calls with a positive delta that occur when the counter is zero must happen before a Wait. Calls with a negative delta, or calls with a positive delta that start when the counter is greater than zero, may happen at any time. Typically this means the calls to Add should execute before the statement creating the goroutine or other event to be waited for. 
https://golang.org/pkg/sync/#WaitGroup.Add